### ISSUE TITLE
Silent empty backups

### ISSUE THREAD

**Issue by Eeems** [g3492/vzborg#13]
If vzdump fails for some reason (No idea why, but an update/reboot fixed it for me, it just kept on saying it got an invalid control message), vzborg will create an empty backup instead of failing.

This is an issue for me as I automatically rotate my backups, which left me without backups for my VMs/Containers on one of my hosts until I noticed it was failing.

Ideally vzborg should capture/log the error and then return an error code when done.

**Comment by g3492** [comment]
Hi Nathaniel. Can you help me with more information, or a way to reproduce the problem? Exactly which was the error in vzdump? Do you have a log? Thanks.

**Comment by Eeems** [comment]
The only error message I was getting was that it was getting an invalid control message. No other information was being provided and I can't replicate it anymore now that I've rebooted the node. I believe that vzdump was thinking that one of my storage locations (cifs share) wasn't mounted, even though it was.

**Comment by g3492** [comment]
OK, thanks. I will try to reproduce a similar problem, and let you know.

**Comment by Eeems** [comment]
Just had it happen again. Looks like it happened 3 days ago:
![image](https://user-images.githubusercontent.com/1176979/132548110-03b194a3-b12d-4c42-b05d-49e4464d12ea.png)
It also doesn't give much in the logs at all:
![image](https://user-images.githubusercontent.com/1176979/132548194-e02eaf9b-ac2e-4b39-ae08-650bf65708e1.png)
![image](https://user-images.githubusercontent.com/1176979/132548273-a92e7a98-c799-4e87-b4a2-3406d719090b.png)

That is the only output of vzdump. Looks like all my nightly backups also have the same error. I'm going to reboot my host as I don't want to go too long without backups. At least I still have local backups from when this started, but my offsites are gone due to vzborg not detecting that it failed and creating the empty backup files.

Anything you think I should check next time I see this happen?

**Comment by Eeems** [comment]
Looks like you can also replicate an empty backup by having vzdump timeout waiting for the lock:
![image](https://user-images.githubusercontent.com/1176979/132705739-69f94fd8-f4b7-4880-914f-47213f51166e.png)

**Comment by Eeems** [comment]
@g3492 poke. This issue is currently making me very nervous about my backup solution.

**Comment by jopicap** [comment]
This is a matter of vzdump not vzborg. When vzdump starts, it tries to `flock` on /var/run/vzdump.lock (semaphore). No second vzdump can start until first finishes. This is coded in vzdump. You can set 'lockwait' in /etc/vzdump.conf to increase from default 180 min to higher value. Check running processes with "ps -ef" and the locking process with "sudo lslocks".

From vzdump source:
`getlock($maxwait); # only one process allowed
`

Regards, JP

**Comment by Eeems** [comment]
No, the bug I'm reporting is that vzborg will upload an empty backup instead of erroring out if vzdump fails. The easiest way to replicate is with timeouts on the lock, but it will happen for any other reason that vzdump fails.

**Comment by jopicap** [comment]
Hmpf. Sorry. Haven't understood the whole problem. Read only partly.

Man page https://pve.proxmox.com/pve-docs/vzdump.1.html does not say a word about command's return-code.
https://borgbackup.readthedocs.io/en/stable/usage/create.html has section **Reading from stdin** stating:

> `... piping to borg creates an archive even if the command piping to borg exits with a failure. In this case, one can end up with truncated output being backed up. 
> `

So, instead of using a pipe
`vzdump ${VM_ID} ... --stdout | borg create` 
is is recommended to use **--content-from-command** option:
`borg create --content-from-command REPO::ARCHIVE -- vzdump --id myvm --stdout`

Maybe something like that. Don't know if the "|& tee" will work:

```
borg create ${borg_extra} --content-from-command --files-cache=disabled --stdin-name=vzborgdump --chunker-params ${default_chunker} -s "${backup_compression}" "${backup_comment}" \
             ${vzborg_repo}::${backup_name} -- vzdump ${VM_ID} --dumpdir ${temp_dir} --quiet ${vzdump_quiet} --compress 0 --mode ${vzdump_mode} --stdout |& tee -a ${detail_file};

```

**Comment by Eeems** [comment]
It looks like `--content-from-command` wasn't added in borgbackup `1.1.16`.

**Comment by Eeems** [comment]
It looks like borg `1.2.4` is now available in the repos and that means `--content-from-command` is now available for use. I'll have to try this out now.

**Comment by Eeems** [comment]
```
2024-10-06 14:22:43 INFO: tar: /tmp/vzdump-lxc-110-2024_10_06-14_22_36.tmp: Cannot open: Permission denied
2024-10-06 14:22:43 INFO: tar: Error is not recoverable: exiting now
```

Something isn't right here, as vzdump can run happily when run by hand with the same options: `vzdump 110 --dumpdir /tmp --quiet 0 --compress 0 --mode snapshot --stdout > /dev/null`
